File: backend/mcp-code-intelligence-python/src/mcp_code_intel/analyzers/graph_analysis/utils/graph_loader.py
"""KSA-163: Graph Loader — Loads subgraphs from relationships table."""
from __future__ import annotations
import sqlite3
from .. import AdjacencyList, SymbolInfo


class GraphLoader:
    def __init__(self, conn: sqlite3.Connection) -> None:
        self._conn = conn

    def load_dependency_graph(self, module: str | None = None) -> AdjacencyList:
        sql = ("SELECT source_symbol_id, target_symbol_id FROM relationships "
               "WHERE kind = 'imports' AND target_symbol_id IS NOT NULL "
               "AND file_path NOT LIKE '%node_modules%' AND file_path NOT LIKE '%vendor%'")
        params: list = []
        if module:
            sql += " AND file_path LIKE ?"; params.append(f"%{module}%")
        return self._load_graph(sql, params)

    def load_reverse_call_graph(self, module: str | None = None) -> AdjacencyList:
        sql = ("SELECT source_symbol_id, target_symbol_id FROM relationships "
               "WHERE kind = 'calls' AND target_symbol_id IS NOT NULL")
        params: list = []
        if module:
            sql += " AND file_path LIKE ?"; params.append(f"%{module}%")
        graph: AdjacencyList = {}
        for row in self._conn.execute(sql, params).fetchall():
            src, tgt = row[0], row[1]
            graph.setdefault(tgt, []).append(src)  # reversed
            graph.setdefault(src, [])
        return graph

    def get_symbol_info(self, symbol_id: int) -> SymbolInfo | None:
        row = self._conn.execute(
            "SELECT s.id, s.name, s.kind, f.relative_path FROM symbols s "
            "JOIN files f ON f.id = s.file_id WHERE s.id = ?", (symbol_id,),
        ).fetchone()
        return SymbolInfo(row[0], row[1], row[2], row[3]) if row else None

    def get_symbol_info_batch(self, ids: list[int]) -> dict[int, SymbolInfo]:
        if not ids: return {}
        ph = ",".join("?" * len(ids))
        rows = self._conn.execute(
            f"SELECT s.id, s.name, s.kind, f.relative_path FROM symbols s "
            f"JOIN files f ON f.id = s.file_id WHERE s.id IN ({ph})", ids,
        ).fetchall()
        return {r[0]: SymbolInfo(r[0], r[1], r[2], r[3]) for r in rows}

    def resolve_symbol_id(self, name: str, file_path: str | None = None) -> int | None:
        sql = "SELECT s.id FROM symbols s JOIN files f ON f.id = s.file_id WHERE s.name = ?"
        params: list = [name]
        if file_path:
            sql += " AND f.relative_path LIKE ?"; params.append(f"%{file_path}%")
        sql += " LIMIT 1"
        row = self._conn.execute(sql, params).fetchone()
        return row[0] if row else None

    def _load_graph(self, sql: str, params: list) -> AdjacencyList:
        graph: AdjacencyList = {}
        for row in self._conn.execute(sql, params).fetchall():
            src, tgt = row[0], row[1]
            graph.setdefault(src, []).append(tgt)
            graph.setdefault(tgt, [])
        return graph

File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/graphanalysis/utils/GraphLoader.kt
/**
 * KSA-163: Graph Loader — Loads subgraphs from the relationships table.
 */
package com.codeintel.analyzers.graphanalysis.utils

import com.codeintel.analyzers.graphanalysis.AdjacencyList
import com.codeintel.analyzers.graphanalysis.SymbolInfo
import java.sql.Connection

class GraphLoader(private val conn: Connection) {

    fun loadDependencyGraph(module: String? = null): AdjacencyList {
        val sql = buildString {
            append("SELECT source_symbol_id, target_symbol_id FROM relationships ")
            append("WHERE kind = 'imports' AND target_symbol_id IS NOT NULL ")
            append("AND file_path NOT LIKE '%node_modules%' AND file_path NOT LIKE '%vendor%' ")
            if (module != null) append("AND file_path LIKE ? ")
        }
        return loadGraph(sql, module)
    }

    fun loadCallGraph(module: String? = null): AdjacencyList {
        val sql = buildString {
            append("SELECT source_symbol_id, target_symbol_id FROM relationships ")
            append("WHERE kind = 'calls' AND target_symbol_id IS NOT NULL ")
            if (module != null) append("AND file_path LIKE ? ")
        }
        return loadGraph(sql, module)
    }

    fun loadReverseCallGraph(module: String? = null): AdjacencyList {
        val sql = buildString {
            append("SELECT source_symbol_id, target_symbol_id FROM relationships ")
            append("WHERE kind = 'calls' AND target_symbol_id IS NOT NULL ")
            if (module != null) append("AND file_path LIKE ? ")
        }
        val graph: AdjacencyList = mutableMapOf()
        conn.prepareStatement(sql).use { stmt ->
            module?.let { stmt.setString(1, "%$it%") }
            val rs = stmt.executeQuery()
            while (rs.next()) {
                val src = rs.getInt("source_symbol_id")
                val tgt = rs.getInt("target_symbol_id")
                graph.getOrPut(tgt) { mutableListOf() }.add(src) // reversed
                graph.getOrPut(src) { mutableListOf() }
            }
        }
        return graph
    }

    fun getSymbolInfo(symbolId: Int): SymbolInfo? {
        val sql = """
            SELECT s.id, s.name, s.kind, f.relative_path as filePath
            FROM symbols s JOIN files f ON f.id = s.file_id WHERE s.id = ?
        """.trimIndent()
        conn.prepareStatement(sql).use { stmt ->
            stmt.setInt(1, symbolId)
            val rs = stmt.executeQuery()
            if (!rs.next()) return null
            return SymbolInfo(rs.getInt("id"), rs.getString("name"), rs.getString("kind"), rs.getString("filePath"))
        }
    }

    fun getSymbolInfoBatch(ids: List<Int>): Map<Int, SymbolInfo> {
        if (ids.isEmpty()) return emptyMap()
        val ph = ids.joinToString(",") { "?" }
        val sql = "SELECT s.id, s.name, s.kind, f.relative_path as filePath FROM symbols s JOIN files f ON f.id = s.file_id WHERE s.id IN ($ph)"
        val map = mutableMapOf<Int, SymbolInfo>()
        conn.prepareStatement(sql).use { stmt ->
            ids.forEachIndexed { i, id -> stmt.setInt(i + 1, id) }
            val rs = stmt.executeQuery()
            while (rs.next()) {
                val info = SymbolInfo(rs.getInt("id"), rs.getString("name"), rs.getString("kind"), rs.getString("filePath"))
                map[info.id] = info
            }
        }
        return map
    }

    fun resolveSymbolId(name: String, filePath: String? = null): Int? {
        val sql = buildString {
            append("SELECT s.id FROM symbols s JOIN files f ON f.id = s.file_id WHERE s.name = ? ")
            if (filePath != null) append("AND f.relative_path LIKE ? ")
            append("LIMIT 1")
        }
        conn.prepareStatement(sql).use { stmt ->
            stmt.setString(1, name)
            filePath?.let { stmt.setString(2, "%$it%") }
            val rs = stmt.executeQuery()
            return if (rs.next()) rs.getInt("id") else null
        }
    }

    private fun loadGraph(sql: String, module: String?): AdjacencyList {
        val graph: AdjacencyList = mutableMapOf()
        conn.prepareStatement(sql).use { stmt ->
            module?.let { stmt.setString(1, "%$it%") }
            val rs = stmt.executeQuery()
            while (rs.next()) {
                val src = rs.getInt("source_symbol_id")
                val tgt = rs.getInt("target_symbol_id")
                graph.getOrPut(src) { mutableListOf() }.add(tgt)
                graph.getOrPut(tgt) { mutableListOf() }
            }
        }
        return graph
    }
}
